File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        max_length = 0
        l = 0  # Left pointer
        
        for h in range(len(s)):  # Right pointer
            while s[h] in char_set:
                char_set.remove(s[l])
                l += 1  # Shrink window from left
            
            char_set.add(s[h])  # Add current char to set
            max_length = max(max_length, h - l + 1)
        
        return max_length

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0, l = 0, h = 0;
        int[] occ = new int[95];
        while (h < s.length() && l <= h) {
            char i = s.charAt(h);
            while (l < h && occ[i - ' '] != 0) {
                char j = s.charAt(l);
                occ[j - ' '] = 0;
                l++;
            }
            occ[i - ' '] = 1;
            max = Math.max(max, h - l + 1);
            h++;
        }
        return max;
    }
}